src/compiler/checker.ts

```diff
@@ -2552,12 +2474,33 @@ namespace ts {
             return links.target;
         }
 
+        /** Indicates that a symbol directly or indirectly resolves to a type-only import or export. */
+        function getTypeOnlyAliasDeclaration(symbol: Symbol): Identifier | ImportSpecifier | ExportSpecifier | undefined {
+            if (!(symbol.flags & SymbolFlags.Alias)) {
+                return undefined;
+            }
+            const links = getSymbolLinks(symbol);
+            if (links.typeOnlyDeclaration === undefined) {
+                const node = getDeclarationOfAliasSymbol(symbol);
+                if (!node) return Debug.fail();
+                if (isTypeOnlyImportOrExportName(node)) {
+                    links.typeOnlyDeclaration = node;
+                }
+                else {
+                    const alias = getImmediateAliasedSymbol(symbol);
+                    links.typeOnlyDeclaration = false; // Prevents circular resolution possible in error scenarios
+                    links.typeOnlyDeclaration = !!alias && getTypeOnlyAliasDeclaration(alias);
+                }
+            }
+            return links.typeOnlyDeclaration || undefined;
+        }
+
         function markExportAsReferenced(node: ImportEqualsDeclaration | ExportSpecifier) {
             const symbol = getSymbolOfNode(node);
             const target = resolveAlias(symbol);
             if (target) {
                 const markAlias = target === unknownSymbol ||
-                    ((target.flags & SymbolFlags.Value) && !isConstEnumOrConstEnumOnlyModule(target));
+                    ((target.flags & SymbolFlags.Value) && !isConstEnumOrConstEnumOnlyModule(target) && !getTypeOnlyAliasDeclaration(symbol));

```

Comment:
If you wanted to bench the removal of the calls to `getTypeOnlyAliasDeclaration`, you'd also need to `undefined`-out here, too~